network, netstat: Centralize the domain & agent data migration

EdDev opened the pull request:
**What this PR does / why we need it**:

The information collected from the domain and guest-agent is processed
in order to associate the interfaces with their corresponding networks
(when possible).
This step has been done at two locations, at the netstat update method
and at the agent-poller.

This change centralizes the data merging at the netstat update method,
leaving in the guest-agent entries only the original data.

With this change, some behaviors are clearer to observe and tests have
been updated to only input relevant data.

---

For backward compatibility with older virt-launchers, when the
guest-agent (domain status) report is seen, apply this logic when
calculating the `infoSource` field:
- When the domain status `Name` filed is set, it is an old virt-launcher,
  therefore, set nothing (the merge occurred in the virt-launcher,
  therefore we do not know if the source is domain-only or domain+GA).
- When the domain status `Name` field is not set (empty), it is a new
  virt-launcher, therefore it can be safely determined as domain+GA.

The transient cases during upgrade (the control plane components are updated in an async manner) is not handled as the situation converges eventually.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes

**Special notes for your reviewer**:
~~Depends on  #7051 ,  #7058~~
~~Only the last 4 commits are to be reviewed in this PR context.~~

**Release note**:
```release-note
NONE
```

EdDev commented:
/uncc omeryahud stu-gott

EdDev commented:
/test all

EdDev commented:
/sig network

orelmisan commented:
Thanks for the PR.
In the PR description it says that the PR: "Fixes #7051 , #7058", which are PRs themselves.
Is this supposed to go at the "Special notes for your reviewer:" section?

EdDev commented:
> Thanks for the PR.
> In the PR description it says that the PR: "Fixes #7051 , #7058", which are PRs themselves.
> Is this supposed to go to the "Special notes for your reviewer:" section?

It was a glitch. Fixed.

EdDev commented:
/test all

EdDev commented:
/retest

EdDev commented:
/test all

EdDev commented:
/cc @orelmisan @ormergi @maiqueb @oshoval @dteplits

oshoval commented:
Thanks

is it fine that now virt-handler will need to process more info ? (it is small a operation, but for all VMs in the node, and for each update event that reach this function
we have only one virt-handler per node, and lots of virt-launchers

EdDev commented:
> Thanks
> 
> is it fine that now virt-handler will need to process more info ? (it is small a operation, but for all VMs in the node, and for each update event that reach this function we have only one virt-handler per node, and lots of virt-launchers

Maybe you should define what is this "more info" you refer to. The virt-handler is already passing over all the lists, processing the data.

oshoval commented:
> > Thanks
> > is it fine that now virt-handler will need to process more info ? (it is small a operation, but for all VMs in the node, and for each update event that reach this function we have only one virt-handler per node, and lots of virt-launchers
> 
> Maybe you should define what is this "more info" you refer to. The virt-handler is already passing over all the lists, processing the data.

True that it already passes over the info, but we are now adding more.
was referring to `MergeAgentStatusesWithDomainData` which is done now in addition to everything else,
wouldn't it be better that virt-launcher will do everything that it can and virt-handler will just decorate it with the pod ips ?

EdDev commented:
> > Maybe you should define what is this "more info" you refer to. The virt-handler is already passing over all the lists, processing the data.
> 
> True that it already passes over the info, but we are now adding more. was referring to `MergeAgentStatusesWithDomainData` which is done now in addition to everything else,

Maybe you should check again the changes in this PR. It eventually was removed as it is not needed.
If it already processed all lists, adding one or two assignments/conditions changes nothing.

> wouldn't it be better that virt-launcher will do everything that it can and virt-handler will just decorate it with the pod ips ?

If it can do *everything* yes, if it requires partitioning, then no.
Partitioning the reporting is what caused the mess that existed to grow. If there is a CPU/IO penalty it needs proof. I do not see one even in theory.

You could take this one step further and attempt to move this whole processing to the virt-launcher, it sounds to me a good objective.

EdDev commented:
change: Rebase

oshoval commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1675,9 +1675,9 @@ func (l *LibvirtDomainManager) GetGuestInfo() (v1.VirtualMachineInstanceGuestAge
 
 // InterfacesStatus returns the interfaces Guest Agent reported
-func (l *LibvirtDomainManager) InterfacesStatus(domainInterfaces []api.Interface) []api.InterfaceStatus {
+func (l *LibvirtDomainManager) InterfacesStatus() []api.InterfaceStatus {
```
whole func equal simply to `l.agentData.GetInterfaceStatus()` isn't it ?

EdDev replied:
Hmm, right. I should remove the condition.
Done

oshoval commented on pkg/network/vmispec/interface.go:
```diff
@@ -39,6 +39,16 @@ func FilterSRIOVInterfaces(ifaces []v1.Interface) []v1.Interface {
 	return sriovIfaces
 }
 
+func FilterInterfacesSpec(ifaces []v1.Interface, predicate func(i v1.Interface) bool) []v1.Interface {
```
unused ?

EdDev replied:
This should be moved to the `network, netstat: Reorder and clean status update steps` commit.
Done

oshoval reviewed (commented):
reviewed part

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/a68b8cfbeac2ac96d3d2d96a619613e294049b85..926352ef83fd9c2a0628059faf70ac72eca5cabf): Answered review comments.
- Relocate one function addition
- Remove a redundant condition.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -197,3 +198,46 @@ func newVMIIfaceStatusFromGuestAgentData(guestAgentInterface api.InterfaceStatus
+	}
+
+	aliasesCoveredByAgent := []string{}
+	for i, interfaceStatus := range interfaceStatuses {
```
The commit on which this is happening is just moving the code as is. Following commits will drop this whole merge.
If there is an optimization needed, lets discuss it on the end result, ok?

oshoval reviewed (commented):
in the 3rd commit
```
In case no guest-agent data exists, the IP fields are overridden by the
pod-cache for all interfaces.
```

can you check please if in the transient state before the first guest-agent report,
there is a change in the behavior comparing to before this PR ?

I mean, lets say there is a guest-agent, but in the around 30 seconds before the first report,
the flow is like there is no guest-agent (after all there was no report, or maybe empty report still ?),
i wonder if now due to the logic change, we will start to see ips that we didn't see before (only for the transient state).
in case we do see a change, at least lets known about it and see we are fine with it
(for bridge / masquerade)

EdDev commented:
> in the 3rd commit
> 
> ```
> In case no guest-agent data exists, the IP fields are overridden by the
> pod-cache for all interfaces.
> ```
> 
> can you check please if in the transient state before the first guest-agent report, there is a change in the behavior comparing to before this PR ?
> 
> I mean, lets say there is a guest-agent, but in the around 30 seconds before the first report, the flow is like there is no guest-agent (after all there was no report), i wonder if now due to the logic change, we will start to see ips that we didn't see before that (only for the transient state) and we do see a change, at least lets known about it and see we are fine with it (for bridge / masquerade)

The unit tests have been adjusted to represent the expected behavior before this PR. If I am not mistaken, no changes to the expected results have been touched in these changes, therefore, I do not expect any change in the behavior.

If you see a missing scenario, then we better add it to the unit tests.

Regarding the transient state, I do not understand what you are suggesting to check... that in this transient period some value appears in the report? Would we really want to add such an e2e test check?
If I am not mistaken, there are some tests that check what is seen when there is no guest-agent. So how is that different?
Anyway, if you think we need to add another e2e test to check something, please propose it.

From my point of view, during the period where there is no guest-agent data at all, the reporting logic behaves the same for guests with or without guest-agent. When the guest-agent data is received and the virt-handler is "seeing" it, then it behaves as if there is an active guest-agent. We have these scenarios covered in the unit tests and AFAIK in the existing e2e test.
I have not managed to detect any edge case we are not covering, but if there is one, please do point it out.

oshoval commented:
> From my point of view, during the period where there is no guest-agent data at all, the reporting logic behaves the same for guests with or without guest-agent. When the guest-agent data is received and the virt-handler is "seeing" it, then it behaves as if there is an active guest-agent. We have these scenarios covered in the unit tests and AFAIK in the existing e2e test. I have not managed to detect any edge case we are not covering, but if there is one, please do point it out.

It seems we are covered, but I personally think that it worth safe, and easiest to check manually one time only about the transient state with a real VM and see if there is a behavior change.
In case there is, we will know about it, and then can act / add test for it.

Beside that lgtm
Thanks

EdDev commented on pkg/virt-launcher/virtwrap/cmd-server/server.go:
```diff
@@ -367,7 +367,7 @@ func (l *Launcher) GetDomain(_ context.Context, _ *cmdv1.EmptyRequest) (*cmdv1.D
 			domainObj.Status.OSInfo = *osInfo
 		}
-		if interfaces := l.domainManager.InterfacesStatus(domainObj.Spec.Devices.Interfaces); interfaces != nil {
+		if interfaces := l.domainManager.InterfacesStatus(); interfaces != nil {
```
I do not understand how the merger is relevant here. Looks to me like it should have been like that in the first place.
The domain is coming from `ListAllDomains` which comes with nothing populated in the status in the first place.

I prefer not to touch this at all in this context. We can change it in a following PR.

EdDev commented:
/retest

EdDev commented:
change: Rebase

EdDev commented:
change: Rebase

AlonaKaplan reviewed (commented):
I suspect that the change may have backward compatibility issues.
Please add to the PR message a section explaining why the change is backward compatible.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -75,40 +75,45 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 			continue
 		}
+
 		podIface, err := c.getPodInterfacefromFileCache(vmi, iface.Name)
```
If I am not mistaken, calling `netvmispec.LookupInterfaceStatusByName(..)` will return a `nil` for SR-IOV interfaces (as they will not appear in the status in case of no-ga).

But most importantly, we should have this covered in the unit tests. If we do not, we should fix it.

EdDev commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1725,65 +1724,14 @@ var _ = Describe("Manager", func() {
-				Expect(libvirtmanager.InterfacesStatus(fakeDomInterfaces)).To(BeNil())
-			})
-
-			It("should return merged list when interfaces exists on both the cache and argument", func() {
```
How about we drop all this context ("on call to InterfacesStatus")?
There is no logic involved, it acts only as a simple proxy.
We do cover this at https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go#L124 .

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -167,20 +169,20 @@ func ifacesStatusFromGuestAgent(vmiIfacesStatus []v1.VirtualMachineInstanceNetwo
 	for _, guestAgentInterface := range guestAgentInterfaces {
 		if vmiIfaceStatus := netvmispec.LookupInterfaceStatusByMac(vmiIfacesStatus, guestAgentInterface.Mac); vmiIfaceStatus != nil {
 			updateVMIIfaceStatusWithGuestAgentData(vmiIfaceStatus, guestAgentInterface, vmiInterfacesSpecByName)
+			vmiIfaceStatus.InfoSource = netvmispec.InfoSourceDomainAndGA
```
The backward comparability with older virt-launchers seems to be here.
Suggested fix is to condition this per the `Name` field value, from the `guestAgentInterface`.

Older virt-launchers will have this non-empty (due to the old merge).
New ones will have nothing here, as `Name` is no longer in use.

* Note: The `InfoSource` exists for only a short period, therefore it is suggested to ignore the case where it was available and not treat it as a special case.

EdDev replied:
End result should be:
- For new launchers, set the source.
- For old launcher, set it to empty (as we do not know the origin really).

AlonaKaplan replied:
The name will be there since you still call the merge internally. I think the merge is redundant now.
EDIT: ignore, I see it is done in the next commit.
The backward compatibility plan sounds good.

AlonaKaplan commented on pkg/network/setup/netstat.go:
```diff
@@ -202,47 +204,19 @@ func newVMIIfaceStatusFromGuestAgentData(guestAgentInterface api.InterfaceStatus
 
-// MergeAgentStatusesWithDomainData merge QEMU interfaces with agent interfaces
-func MergeAgentStatusesWithDomainData(domInterfaces []api.Interface, interfaceStatuses []api.InterfaceStatus) []api.InterfaceStatus {
+// mergeAgentStatusesWithDomainData merge QEMU interfaces with agent interfaces
```
Why the merge method is still needed? Couldn't find a place using the `name` field that is added here.

EDIT: ignore, I see it is done in the next commit.

AlonaKaplan commented on pkg/network/setup/netstat.go:
```diff
@@ -79,40 +79,45 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
+		domain.Status.Interfaces = mergeAgentStatusesWithDomainData(domain.Spec.Devices.Interfaces, domain.Status.Interfaces)
 		interfacesStatus = ifacesStatusFromGuestAgent(interfacesStatus, vmi.Spec.Domain.Devices.Interfaces, domain.Status.Interfaces)
+
+		ifacesSpec := netvmispec.FilterInterfacesSpec(vmi.Spec.Domain.Devices.Interfaces, func(i v1.Interface) bool {
```
Instead of `ifacesSpec` I would call it `forwardingIfacesSpec` or maybe `natIfacesSpec`.

AlonaKaplan replied:
slirp is NAT-ed
> User Networking is implemented using "slirp", which provides a full TCP/IP stack within QEMU and uses that stack to > implement a virtual NAT'd network.
https://wiki.qemu.org/Documentation/Networking#User_Networking_.28SLIRP.29

EdDev replied:
Hmm.. I meant it is not NAT-ed by kubevirt. The binding itself is not defining a NAT for SLIRP, it treats it as an "application".
For me, it will be confusing, but if you think it is fine, I really do not mind.

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/6cedd2833863371b33d37b901e46192c244b1f01..7ca3ed84d217cfb0f5e7ce632c3f9718b337f4fa): Added backward compatibility handling and kept the domain status `Name` field.

EdDev commented:
> I suspect that the change may have backward compatibility issues. Please add to the PR message a section explaining why the change is backward compatible.

Done
Added a dedicated commit for this and mentioned it on the PR description.

AlonaKaplan commented on pkg/network/setup/netstat.go:
```diff
@@ -167,20 +169,20 @@ func ifacesStatusFromGuestAgent(vmiIfacesStatus []v1.VirtualMachineInstanceNetwo
 	ifaceStatus.InterfaceName = guestAgentIface.InterfaceName
-	ifaceStatus.InfoSource = guestAgentIface.InfoSource
 
-	// Update IP/s only if the interface has no Masquerade/Slirp binding
```
I don't insist:)

AlonaKaplan commented on pkg/network/setup/netstat.go:
```diff
@@ -168,7 +168,16 @@ func ifacesStatusFromGuestAgent(vmiIfacesStatus []v1.VirtualMachineInstanceNetwo
 		if vmiIfaceStatus := netvmispec.LookupInterfaceStatusByMac(vmiIfacesStatus, guestAgentInterface.Mac); vmiIfaceStatus != nil {
 			updateVMIIfaceStatusWithGuestAgentData(vmiIfaceStatus, guestAgentInterface, vmiInterfacesSpecByName)
-			vmiIfaceStatus.InfoSource = netvmispec.InfoSourceDomainAndGA
+			// For backward compatability with older virt-launchers, apply this logic:
```
In case the `name` is there, cannot you check the `intefaceName` is there as well?
Anything reported from the GA should have `interfaceName` isn't it?
Maybe it is even enough to check just the `interfaceName` then the `name` can be dropped.

EdDev replied:
I think you are right. We can use `interfaceName` to determine if it came from the GA or not.
There should be no case where this field is not populated when originated for the agent.

I will re-do this, thanks!

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/7ca3ed84d217cfb0f5e7ce632c3f9718b337f4fa..9c205a8be7deeca57440d09a04573bef996c19c6): Added an unit test, changed a variable name and changed the backward compatibility algorithm (which allowed the `Name` field to be removed).

AlonaKaplan commented on pkg/network/setup/netstat.go:
```diff
@@ -168,7 +168,14 @@ func ifacesStatusFromGuestAgent(vmiIfacesStatus []v1.VirtualMachineInstanceNetwo
+			// - When the domain status `InterfaceName` field is not set (empty), the data originates from a merge
+			//   at an old virt-launcher and comes from the domain spec. In such cases, no action is to be taken.
+			if guestAgentInterface.InterfaceName != "" {
+				vmiIfaceStatus.InfoSource = netvmispec.InfoSourceDomainAndGA
```
I think `updateVMIIfaceStatusWithGuestAgentData` should be moved here as well.

EdDev replied:
That function updates the VMI status with the GA ifaceName and IP/s. Why would this be wrong when treating old vmi-launchers?
For data that originated from GA report, it will be correct.
For data that originated from the domain-only (in the merge), it will also be correct (as this data is not available from the domain).

If we need to compliment the data later on, it will be done later from reading from the podCache.

AlonaKaplan replied:
I understand that currently it works as is. But `updateVMIIfaceStatusWithGuestAgentData` is designed to deal with GA data only. For example, it can assume `InterfaceName` is always there. The fact that currently it works on a "fake" GA data doesn't mean it will be the case in the future and I prefer not to maintain this.

But it is not a blocker for me, so if I didn't convince you, you can keep it as is.

EdDev replied:
What I am trying to say is that it works fine for both new and old launchers, just because the data there can only come from the GA. If it did not come from it, it will be empty and it is still correct (because no one else will fill it up until this step).

But when you pointed this function, I realized it can be simplified and remove a condition.
I am pushing a change to simplify it even further.

oshoval commented:
> [change](https://github.com/kubevirt/kubevirt/compare/9c205a8be7deeca57440d09a04573bef996c19c6..546bdceba24a5f6a03716a079f32d5f7c47662f2): There is no need to condition the update of the VMI interface status by the guest-agent data. One step later, it will be overridden by the pod cache data in case these are SR-IOV or SLIRP.

Ain't it better that it will be explicit like it was before ?
what about bridge btw, same ? (unless not related here)

AlonaKaplan commented:
/approve

EdDev commented:
> There is no need to condition the update of the VMI interface status by the guest-agent data. One step later, it will be overridden by the pod cache data in case these are SR-IOV or SLIRP.

> Ain't it better that it will be explicit like it was before ?

Why is it better? I do not think a condition is good, especially if we need to ask the same thing twice.
But feel free to reason for keeping it.

> what about bridge btw, same ? (unless not related here)

What bridge? I do not understand the scenario you are trying to describe.
If it is related or not, you know. I don't understand what you are asking or with what scenarios you see a problem.

oshoval commented on pkg/network/setup/netstat.go:
```diff
@@ -139,68 +144,64 @@ func vmiInterfaceKey(vmiUID types.UID, interfaceName string) string {
 		if vmiIfaceStatus := netvmispec.LookupInterfaceStatusByMac(vmiIfacesStatus, guestAgentInterface.Mac); vmiIfaceStatus != nil {
-			updateVMIIfaceStatusWithGuestAgentData(vmiIfaceStatus, guestAgentInterface, vmiInterfacesSpecByName)
+			updateVMIIfaceStatusWithGuestAgentData(vmiIfaceStatus, guestAgentInterface)
+			// For backward compatability with older virt-launchers, apply this logic:
+			// - When the domain status `InterfaceName` field is set, the data originates from the guest-agent.
+			//   This is true for old virt-launchers and new ones alike.
+			// - When the domain status `InterfaceName` field is not set (empty), the data originates from a merge
+			//   at an old virt-launcher and comes from the domain spec. In such cases, no action is to be taken.
+			if guestAgentInterface.InterfaceName != "" {
+				vmiIfaceStatus.InfoSource = netvmispec.InfoSourceDomainAndGA
```
worth to have it in a function ?
we might remove it in some point in the future, so it will be self contained

EdDev replied:
The only thing that we may remove in the future is the condition. The comment is explaining it, not the assignment.
So if you we move it to a function, it has a potential to just wrap the one-line assignment, which is very odd.

But I usually judge this per the proposed name. I do not have a good name that can embed in it the condition and assignment for this specific case. If you have a good name to describe it, and not change this name when this future you mentioned will come, then we can consider exporting this to a function.

oshoval replied:
what about `if ! isOutdatedVirtLauncher` ?
```
func isOutdatedVirtLauncher(....) bool {
  return guestAgentInterface.InterfaceName == ""
}
```

And then we will be able to just remove the condition and the function (as now, but it is self contained)
maybe it will allow to move the reasoning comment to the function ? (and if we want to be even more explicit, to explain that only the condition can be removed in the future as you explained)

EdDev replied:
I'm fine to export the condition with the comment to a function.
@AlonaKaplan , are you ok with this?

EdDev replied:
Done

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/546bdceba24a5f6a03716a079f32d5f7c47662f2..e605b5892e7d18a8b45677020c4fc396f4beb927): Extracted the backward compatibility condition to a function.

oshoval reviewed (approved):
Thanks

EdDev commented:
/retest

EdDev commented:
/retest

EdDev commented:
/retest
